### PR TITLE
Initial PR for Ensemble Selection

### PR DESCRIPTION
**Pull request by franchuterivera**
Ideally, we would like to have the full ensemble building process as a common package.

Ensemble builder requires logging, metrics, and other shared collaterals to be enabled first.

This PR is a proof of concept to enable ensemble selection as a commonly shared package for AutoML frameworks.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+repos:
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v0.761
+    hooks:
+      - id: mypy
```
Comment (mfeurer) [inline .pre-commit-config.yaml:5]: Hey, just checking - this now checks all python files with mypy?

Comment (mfeurer) [reply .pre-commit-config.yaml:5]: Also, once we know the overall setup is working, can we make this as strict as possible, i.e. no `Any` and requiring everything to be typed?

File: .github/workflows/pytest.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
+name: Tests
+
+on: [push, pull_request]
+
+jobs:
+  ubuntu:
+
+    runs-on: ubuntu-18.04
```
Comment (mfeurer) [inline .github/workflows/pytest.yml:8]: Now that this doesn't require the pyrfr (and the pyrfr being available as a wheel), can we relax this?

File: .github/workflows/pytest.yml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    - uses: actions/checkout@v2
+    - name: Setup Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Conda Install test dependencies
+      if: matrix.use-conda == true
+      run: |
+        # Miniconda is available in $CONDA env var
+        $CONDA/bin/conda create -n testenv --yes pip wheel gxx_linux-64 gcc_linux-64 swig python=${{ matrix.python-version }}
```
Comment (mfeurer) [inline .github/workflows/pytest.yml:30]: We don't need the compilers and swig here, right?

Comment (mfeurer) [reply .github/workflows/pytest.yml:30]: (and we might also no longer need them for Auto-sklearn, right?)

File: .github/workflows/pytest.yml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      if: matrix.use-conda == false
+      run: |
+        python -m pip install --upgrade pip
+        if [[ `python -c 'import platform; print(platform.python_version())' | cut -d '.' -f 2` -eq 6 ]]; then
+          # Numpy 1.20 dropped suppert for Python3.6
+          pip install "numpy<=1.19"
+        fi
+        pip install -r requirements.txt
+        sudo apt-get update
+        sudo apt-get remove swig
```
Comment (mfeurer) [inline .github/workflows/pytest.yml:43]: this can also be removed I guess?

File: shared/ensemble_building/ensemble_selection.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+import numpy as np
+
+from sklearn.pipeline import Pipeline
+
+from shared.ensemble_building.abstract_ensemble import AbstractEnsemble
+
+
+class EnsembleSelection(AbstractEnsemble):
+    """Implementation of Ensemble Selection from the paper:
+    Ensemble selection from libraries of models, Caruana 2016.
```
Comment (mfeurer) [inline shared/ensemble_building/ensemble_selection.py:14]: Could you please give the full reference?

File: shared/ensemble_building/ensemble_selection.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        The maximum number of individual members that an ensemble can have.
+
+    calculate_loss: Callable
+        A function with signature calculate_loss(solutions, predictions) that estimate
+        the loss of the current predictions vs the expected solutions. The current
+        implementation of ensemble selections solves a minimization problem.
+    random_state: np.random.RandomState
+        A random number generator used to break ties when 2 models provide the same
+        performance gain.
+    bagging: bool
```
Comment (mfeurer) [inline shared/ensemble_building/ensemble_selection.py:31]: We can probably drop old and broken code now?

File: shared/ensemble_building/ensemble_selection.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+        Enables the bagging approach from Ensemble selection from libraries
+        of models, Caruana 2016.
+    mode: str ['fast', 'slow']
```
Comment (mfeurer) [inline shared/ensemble_building/ensemble_selection.py:34]: Same here?

File: shared/ensemble_building/ensemble_selection.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+        selection
+
+    Returns
+    -------
+    self
+    """
+    def __init__(
+        self,
+        ensemble_size: int,
+        calculate_loss: Callable,
```
Comment (mfeurer) [inline shared/ensemble_building/ensemble_selection.py:45]: Will this be updated once the metric module is copied, too?

Comment (mfeurer) [reply shared/ensemble_building/ensemble_selection.py:45]: Maybe it would be good to also already initialize the metric module so all required dependencies are already there and this initial PR can be merged in one go?

**Review by mfeurer — COMMENTED** [review]
This looks good overall. Just two questions (besides the comments I left):
1. Is there a reason you named the package `shared` and not `commons` or `common` as the repository?
2. Where in the dependency tree would this be included? Would it live under `autosklearn`? If yes, the imports won't work and you need to use relative imports inside the package.
